### PR TITLE
Clarified difference between FIFOQueue and StagingArea

### PR DESCRIPTION
**Pull request by ashwhall**
Clarified that StagingArea doesn't guarantee ordered delivery, contrasting tf.FIFOQueue

### CODE REVIEW

**Comment by ashwhall** [comment]
I signed it!

File: tensorflow/docs_src/performance/performance_models.md [hunk 1]
```diff
@@ -29,12 +29,12 @@ implementation is made up of 3 stages:
...
-similar to @{tf.FIFOQueue}. The difference is that `StagingArea` offers simpler
-functionality and can be executed on both CPU and GPU in parallel with other
-stages. Breaking the input pipeline into 3 stages that operate independently in
-parallel is scalable and takes full advantage of large multi-core environments.
-The rest of this section details the stages followed by details about using
-`data_flow_ops.StagingArea`.
+similar to @{tf.FIFOQueue}. The difference is that `StagingArea`  does not 
+guarantee FIFO ordering, but offers simpler functionality and can be executed 
+on both CPU and GPU in parallel with other stages. Breaking the input pipeline
+into 3 stages that operate independently inparallel is scalable and takes full
```
Comment (yifeif) [inline tensorflow/docs_src/performance/performance_models.md]: space between inparallel.

**Comment by sjperkins** [comment]
I believe that StagingArea does behave in a FIFO manner. Internally, StagingArea uses a C++ [dequeue](https://github.com/tensorflow/tensorflow/blob/235192d47cfb375c0cc93c1deefb9e440715bf35/tensorflow/core/kernels/stage_op.cc#L46). 

[Puts](https://github.com/tensorflow/tensorflow/blob/235192d47cfb375c0cc93c1deefb9e440715bf35/tensorflow/core/kernels/stage_op.cc#L115) push data onto the back of the deque, while [Gets](https://github.com/tensorflow/tensorflow/blob/235192d47cfb375c0cc93c1deefb9e440715bf35/tensorflow/core/kernels/stage_op.cc#L134) pop data off the front.

**Comment by ashwhall** [comment]
I've observed different behaviour, and the documentation says that ordering isn't guaranteed.

**Comment by sjperkins** [comment]
The [test cases](https://github.com/tensorflow/tensorflow/blob/235192d47cfb375c0cc93c1deefb9e440715bf35/tensorflow/python/kernel_tests/stage_op_test.py#L240-L275) assume FIFO behaviour.

Are you extracting values in different threads?

**Comment by ashwhall** [comment]
Yes I have two filling the StagingAreas. One for training set and one for validation set, and also a thread running a validation pass every five seconds.

**Comment by drpngx** [comment]
@yifeif good to go?

Jenkins, test this please.

**Comment by sb2nov** [comment]
@yifeif ping ?

**Review by yifeif — APPROVED** [review]

File: tensorflow/docs_src/performance/performance_models.md (hunk 1 above)
Comment (yifeif) [reply tensorflow/docs_src/performance/performance_models.md]: Actually, still need a space here.

**Comment by Remember2018** [comment]
Hi, I have the following code to be feed into the graph.

    data_iter = get_train_dataflow(batch_size=cfg.TRAIN.BATCH_SIZE_PER_GPU)
    ds = tf.data.Dataset()
    ds = ds.from_generator(lambda: map(lambda x: tuple(x.values()), data_iter),
                           (tf.float32, tf.float32, tf.int64, tf.int32),
                           (tf.TensorShape([None, None, None, 3]),  # images
                            tf.TensorShape([None, None, 4]),  # gt_boxes
                            tf.TensorShape([None, None]),  # gt_labels
                            tf.TensorShape([None, ])))  # orig_gt_counts
    ds = ds.prefetch(buffer_size=256)
    ds = ds.make_one_shot_iterator()
    net_inputs = ds.get_next()
    
    then, the inputs are feed into the `StagingArea`,

    inputs_list = []
    for i in range(num_gpus):
        inputs_list.append(net_inputs)
    put_op_list = []
    get_op_list = []
    for i in range(num_gpus):
        with tf.device("/GPU:{}".format(i)):
            area = StagingArea(dtypes=[tf.float32, tf.float32, tf.int64, tf.int32],
                               shapes=(tf.TensorShape([None, None, None, 3]),
                                       tf.TensorShape([None, None, 4]),
                                       tf.TensorShape([None, None]),
                                       tf.TensorShape([None, ])),
                               capacity=64)
            put_op_list.append(area.put(inputs_list[i]))
            get_op_list.append(area.get())
    
it seems the code is wrong, do I need to generate a `tf.data.DataSet()` for each GPU stage area? The `tf.FIFOQueue` seems not to support the `None` shape of data but fixed known shape of data. But for object detection, the image size are not fixed. Could you please give some advice? Thanks.
